### ISSUE TITLE
What is a reasonable serial default for DataCommunicator functions?

### ISSUE THREAD

**Issue by jcotela** [KratosMultiphysics/Kratos#3646]
The question at hand is what to do with the two-buffer wrappers for MPI calls. Given the debate around this point in #3340 and in the technical committee, I wanted to open a separate issue on this specific topic. As currently implemented in #3340, functions such as
https://github.com/KratosMultiphysics/Kratos/blob/7878279d474f6f9288b098bf0c977988c7782a7b/kratos/includes/data_communicator.h#L307-L311
Are simply empty in the serial version, so the output buffer is left untouched. This is in contrast to the return buffer version
https://github.com/KratosMultiphysics/Kratos/blob/7878279d474f6f9288b098bf0c977988c7782a7b/kratos/includes/data_communicator.h#L284-L287
where something is always returned.

I am not too happy with the current design, mainly because it is inconsistent (functions that do one thing in serial and a different one in parallel). I see different options going forward, but I'd like to hear the opinion of the eventual users of this class and the @KratosMultiphysics/design-committee in particular:
- Leave things as they are now. The behavior is inconsistent, but documented by tests, and my recommendation would be to only use the two-buffer functions within `if (rComm.IsDistributied())` checks.
- Return some reasonable value (for example, for reduction-based methods, the local value) as output. This introduces a subtler problem: for some methods no reasonable return value exists (send-receive if not sending to rank 0, for example). In such cases, it might make sense to return an empty buffer (the return buffer version does so), but this goes against the rule never resizing the buffer, which the mpi version follows strictly.
- Just remove the two-buffer methods (but they are in general more efficient if the user knows what they are doing).
- Return a reasonable value if possible, throw an error otherwise. In practice, this means that functions that don't have an obvious return value in serial can only be used within `if (rComm.IsDistributied())` checks.

What do you think? FYI @KratosMultiphysics/design-committee @RiccardoRossi @pooyan-dadvand @roigcarlo @adityaghantasala.

**Comment by pooyan-dadvand** [comment]
I don't like inconsistencies. Unless a high performance impact justifying it. 

About the send/receive I would just give an error in serial as they should/would not be called in serial because logically you should not enter in the function which needs to transfer data to some neighbours.

**Comment by philbucher** [comment]
the @KratosMultiphysics/design-committee discussed this together with @jcotela and we reached a consensus:
- The methods will behave the same in serial and in MPI => this might imply a small overhead in serial, but we think that a correct behaving code is much more important. I.e. in the `Max`-example from above `rLocalValues` would be copied to `rGlobalValues`
- the C-functions will work with "by-reference" (same as / to be consistent with MPI itself), whereas the python-versions will work with return-values. For python by-reference would not work e.g. for doubles and ints.

guys please feel free to add if I forgeot something

@jcotela told me that he is already implementing this, I guess we will see the final solution soon in a PR

**Comment by jcotela** [comment]
To add to @philbucher's points, the agreement was serial functions would either **act as the MPI version** when possible (if there is a "logical" default behavior through an internal copy, or if the "communication" is between rank 0 and rank 0) **or throw an error otherwise**. This should allow implementing algorithms that have a reasonable serial behavior without having to check for MPI availability.

For more complex cases, one should still rely on `DataCommunicator.IsDistributed()` to check if MPI is available/required.
